frontend/src/metabase/databases/components/DatabaseForm/DatabaseForm.tsx

```diff
@@ -41,6 +43,8 @@ const DatabaseForm = ({
   onSubmit,
   onCancel,
   onEngineChange,
+  // oisin: I don't like drilling this through this component, all input welcome
+  onFormDirty,
```

Comment:
I think the only non-ugly way to do it would be to move the `useBeforeUnload` somewhere down the component tree, so it can access the form context. Would it be useful to have this kind of confirmation during the instance setup flow (we let people connect their first database during the instance setup) ? If so, then it could be the right call

@ranquild would appreciate your opinion here

Reply:
Ah yeah I get what you're saying - the idea of keeping it "high up" on the component tree was to keep what I think are related. We have `useMount` and `handleSubmit` (i.e. dealing with loading and accepting the form) so I felt it made sense to keep `onBeforeUnload` (dealing with unloading the form)  there as well. I just didn't want to throw a `beforeUnload` hook on the `FormBody` since I thought it would be confusing for someone looking at this set of components for the first time (and would also just seem a bit weird to me)

The other option that I am currently looking into is to plant a `isDirty` flag within the Redux state, then we can just dispatch a `setDirty` from the `DatabaseFormBody`, then access it in the `DatabaseEditApp`. That will get rid of the drilling and keep things a little cleaner

Reply:
> Ah yeah I get what you're saying - the idea of keeping it "high up" on the component tree was to keep what I think are related. We have useMount and handleSubmit (i.e. dealing with loading and accepting the form) so I felt it made sense to keep onBeforeUnload (dealing with unloading the form) there as well

Yeah, that makes a ton of sense 👍 

> The other option that I am currently looking into is to plant a isDirty flag within the Redux state, then we can just dispatch a setDirty from the DatabaseFormBody, then access it in the DatabaseEditApp. That will get rid of the drilling and keep things a little cleaner

I think the current approach would work better here. If we move this state to Redux, we'd have two different sources of truth that can get out of sync (actual formik state + redux flag we have to keep in sync)

Reply:
> I think the current approach would work better here. If we move this state to Redux, we'd have two different sources of truth that can get out of sync (actual formik state + redux flag we have to keep in sync)

Yeah, I see what you mean. I will bring back the prop-drilled code for now - maybe the Redux solution would be more useful in a more complex situation than this one